lib/default.nix

```
@@ -71,8 +66,24 @@ let {
     mkfs.${x.format} ${q.device}
   '';
 
-  create.devices = q: x: ''
-    ${concatStrings (mapAttrsToList (name: create-f { device = "/dev/${name}"; }) x.content)}
+  create.devices = q: x: let
+    raid-devices = lib.filterAttrs (_: dev: dev.type == "mdadm") x.content;
+    other-devices = lib.filterAttrs (_: dev: dev.type != "mdadm") x.content;
+  in ''
+    ${concatStrings (mapAttrsToList (name: create-f { device = "/dev/${name}"; }) other-devices)}
+    ${concatStrings (mapAttrsToList (name: create-f { device = "/dev/${name}"; name = name; }) raid-devices)}
+  '';
+
+  create.mdraid = q: x: ''
+    echo ${q.device} >&2
+    RAIDDEVICES_N_${x.name}=$((''${RAIDDEVICES_N_${x.name}:-0}+1))
+    RAIDDEVICES_${x.name}="''${RAIDDEVICES_${x.name}:-}${q.device} "
+  '';
+
+  create.mdadm = q: x: ''
+    echo "y" | mdadm --create /dev/md/${q.name} --level=${toString x.level} --raid-devices=''${RAIDDEVICES_N_${q.name}} ''${RAIDDEVICES_${q.name}}
```

I would say mirror raid is the most common:

```suggestion
    echo "y" | mdadm --create /dev/md/${q.name} --level=${toString x.level or 1} --raid-devices=''${RAIDDEVICES_N_${q.name}} ''${RAIDDEVICES_${q.name}}
```